src/hotspot/share/tsan/tsanOopMap.cpp

```diff
@@ -414,81 +247,86 @@ namespace TsanOopMapImpl {
+                      moves_this_cycle, remaining_moves);
     }
     log_debug(gc)("Tsan: Move %d passes", passes);
   }
+}  // namespace TsanOopMapImpl
 
-  void TsanOopSizeMap::rebuild_oops_map(BoolObjectClosure *is_alive,
-                                        OopClosure *pointer_adjuster) {
-    ResourceMark rm;
-    GrowableArray<PendingMove> moves(MAX2((int)(oop_map->size() / 100),
-                                          100000));
-    bool disjoint_regions;
-    int n_downward_moves;
-    char *min_low, *max_high;
+static OopStorage* _weak_oop_storage;
+static TsanOopMapTable* _oop_map;
 
-    {
-      disjoint_regions = collect_oops(is_alive, pointer_adjuster, &moves,
-                                      &n_downward_moves, &min_low, &max_high);
-    }
-    if (moves.length() == 0) {
-      return;
-    }
+// This is called with TSAN_ONLY, as we want to always create the weak
+// OopStorage so the number matches with the 'weak_count' in oopStorageSet.hpp.
+void TsanOopMap::initialize_map() {
+  // No need to do register_num_dead_callback for concurrent work as we do
+  // TsanOopMapTable cleanup, i.e. removing entries for freed objects during
+  // GC by calling TsanOopMap::notify_tsan_for_freed_and_moved_objects() from
+  // WeakProcessor.
+  _weak_oop_storage = OopStorageSet::create_weak("Tsan weak OopStorage", mtInternal);
+  assert(_weak_oop_storage != NULL, "sanity");
+
+  TSAN_RUNTIME_ONLY(
+    _oop_map = new TsanOopMapTable();
+  );
+}
 
-    // Notifying TSan is straightforward when source and target regions
-    // do not overlap:
-    if (disjoint_regions) {
-      for (int i = 0; i < moves.length(); ++i) {
-        const PendingMove &m = moves.at(i);
-        __tsan_java_move(m.source_begin(), m.target_begin(), m.n_bytes);
-      }
-      return;
-    }
+void TsanOopMap::destroy() {
+  delete _oop_map;
+}
 
-    // Source and target ranges overlap, the moves need to be ordered to prevent
-    // overwriting. Overall, this can take N^2 steps if only one object can be
-    // moved during the array traversal; however, when we are dealing with
-    // compacting garbage collector, observation shows that the overwhelming
-    // majority of the objects move in one direction. If we sort the moves (in
-    // the ascending order if dominant direction is downwards, in the descending
-    // order otherwise), chances are we will be able to order the moves in a few
-    // traversals of the moves array.
-    {
-      moves.sort((2 * n_downward_moves > moves.length()) ? lessThan : moreThan);
-      log_debug(gc)("Tsan: sort %d objects", moves.length());
-    }
+OopStorage* TsanOopMap::oop_storage() {
+  assert(_weak_oop_storage != NULL, "sanity");
+  return _weak_oop_storage;
+}
 
-    {
-      handle_overlapping_moves(moves, min_low, max_high);
-    }
+// Called during GC by WeakProcessor.
+void TsanOopMap::notify_tsan_for_freed_and_moved_objects() {
+  if (_oop_map == nullptr) {
```

Comment:
Could this be an `assert(_oop_map != nullptr)`?
Also, do we want to keep the old `assert(SafepointSynchronize::is_at_safepoint())`?